stub/src/main/java/io/grpc/stub/BlockingClientCall.java

```diff
@@ -0,0 +1,407 @@
+    boolean origWriteClosed = writeClosed;
+    writeClosed = true;
+
+    if (!origWriteClosed && getClosedStatus() == null) {
+      call.halfClose();
+    } else if (origWriteClosed) {
+      throw new IllegalStateException("halfClose cannot be called after stream terminated");
+    }
+
+    executor.add(NoOpRunnable.INSTANCE);
+  }
+
+  /**
+   * Status that server sent when closing channel from its side.
+   *
+   * @return null if stream not closed by server, otherwise Status sent by server
+   */
+  Status getClosedStatus() {
+    drainQuietly();
+    return closedStatus;
+  }
+
+  /**
+   * Check for whether some action is ready.
+   *
+   * @return True if legal to write and writeOrRead can run without blocking
+   */
+  @VisibleForTesting
+  boolean isEitherReadOrWriteReady() {
+    return (isWriteLegal() && isWriteReady()) || isReadReady();
+  }
+
+  /**
+   * Check whether there are any values waiting to be read.
+   *
+   * @return true if read will not block
+   */
+  @VisibleForTesting
+  boolean isReadReady() {
+    drainQuietly();
+
+    return !buffer.isEmpty();
+  }
+
+  /**
+   * Check that write hasn't been marked complete and stream is ready to receive a write (so will
+   * not block).
+   *
+   * @return true if legal to write and write will not block
+   */
+  @VisibleForTesting
+  boolean isWriteReady() {
+    drainQuietly();
+
+    return isWriteLegal() && call.isReady();
+  }
+
+  /**
+   * Check whether we'll ever be able to do writes or should terminate.
+   * @return True if writes haven't been closed and the server hasn't closed the stream
+   */
+  private boolean isWriteLegal() {
+    return !writeClosed && closedStatus == null;
+  }
+
+  /**
+   * calls this.executor's waitAndDrain or waitAndDrainWithTimeout.
+   *
+   * @throws TimeoutException If no events become available on the queue within the specified
+   *     timeout or processing them exceeds the timeout
+   */
+  private void waitAndDrainExecutorOrTimeout(long start, long duration)
+      throws InterruptedException, TimeoutException {
+    long soFar = System.nanoTime() - start;
+    if (soFar >= duration) {
+      throw new TimeoutException();
+    }
+    // Let threadless executor do stuff until there is something for us to check
+    executor.waitAndDrainWithTimeout(duration - soFar);
+
+    if (System.nanoTime() - start > duration) {
+      throw new TimeoutException();
+    }
+  }
+
+  StartableListener<RespT> getListener() {
+    return listener;
+  }
+
+  private void drainQuietly() {
+    try {
+      executor.drain();
+    } catch (InterruptedException e) {
+      logger.warning("Draining interrupted: " + e.getMessage());
+      Thread.currentThread().interrupt();
+    }
+  }
+
+  /**
+   * wake up write if it is blocked.
+   */
+  void handleReady() {
+    executor.add(NoOpRunnable.INSTANCE);
+  }
+
+  private final class QueuingListener extends StartableListener<RespT> {
+
+    private final ClientCall<ReqT, RespT> call;
```

Comment:
Delete `call` and `buffer`, or make this class `static`? I'd favor removing the fields.